src/include/miopen/norm/problem_description.hpp

```diff
@@ -0,0 +1,109 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include <miopen/problem_description_base.hpp>
+#include <miopen/activ.hpp>
+#include <miopen/tensor.hpp>
+
+#include <cassert>
+#include <string>
+
+namespace miopen {
+
+struct NetworkConfig;
+
+namespace norm {
+
+struct ProblemDescription : ProblemDescriptionBase
+{
+    ProblemDescription(miopenLayerNormMode_t mode_,
+                       const TensorDescriptor& xDesc_,
+                       const TensorDescriptor& weightDesc_,
+                       const TensorDescriptor& biasDesc_,
+                       const TensorDescriptor& yDesc_,
+                       const TensorDescriptor& meanDesc_,
+                       const TensorDescriptor& rstdDesc_,
+                       float epsilon_,
+                       int32_t normalized_dim_)
+        : mode(mode_),
+          xDesc(xDesc_),
+          weightDesc(weightDesc_),
+          biasDesc(biasDesc_),
+          yDesc(yDesc_),
+          meanDesc(meanDesc_),
+          rstdDesc(rstdDesc_),
+          epsilon(epsilon_),
+          normalized_dim(normalized_dim_)
+    {
+    }
+
+    miopenLayerNormMode_t GetMode() const { return mode; }
+    const TensorDescriptor& GetXDesc() const { return xDesc; }
+    const TensorDescriptor& GetWeightDesc() const { return weightDesc; }
+    const TensorDescriptor& GetBiasDesc() const { return biasDesc; }
+    const TensorDescriptor& GetYDesc() const { return yDesc; }
+    const TensorDescriptor& GetMeanDesc() const { return meanDesc; }
+    const TensorDescriptor& GetRstdDesc() const { return rstdDesc; }
+    float GetEpsilon() const { return epsilon; }
+    int32_t GetNormalizedDim() const { return normalized_dim; }
+
+    bool IsRank2Dim1() const { return (xDesc.GetLengths().size() == 2) && (normalized_dim == 1); }
+    bool IsRank4Dim1() const { return (xDesc.GetLengths().size() == 4) && (normalized_dim == 1); }
+    bool IsLargeSize() const
+    {
+        auto dims = xDesc.GetLengths();
+
+        size_t outer_size = 1;
+        for(size_t i = 0; i < normalized_dim; i++)
+        {
+            outer_size *= dims[i];
+        }
+
+        return (outer_size > 32);
```

Comment:
[Q] Suspiciously small limit. Do you have 32 bits in mind or limit is really only 32 pixels?

Reply:
This conditional statement was added because the compoasble kernel performs better when outer_size is larger than 32. In fact, it works flawlessly even at larger sizes.

Reply:
Would you like to resolve this in the subsequent PR?

[Notice] Please don't be afraid of ugly code: sometimes we have to write it. Ugly code is a sign that the design needs to be improved. Therefore, it should be in sight, not hidden under the carpet ;)